File: src/sync_plan.rb
# frozen_string_literal: true

require "json"
require_relative "sync_result"

module SyncLabels
  PlanEntry = Struct.new(:action, :name, :desired, :reason, keyword_init: true) do
    def to_h
      value = {
        "action" => action.to_s,
        "name" => name
      }
      value["desired"] = desired unless desired.nil?
      value["reason"] = reason.to_s unless reason.nil?
      value
    end
  end

  class SyncPlan
    ACTION_COUNTS = {
      create: :created,
      update: :updated,
      rename: :renamed,
      delete: :deleted,
      unchanged: :unchanged,
      preserve: :preserved
    }.freeze
    ACTIONS_WITH_DESIRED = %i[create update rename].freeze
    DELETE_REASONS = %i[legacy_alias stale_managed].freeze

    attr_reader :entries, :counts

    def initialize(entries:)
      raise ArgumentError, "同步计划 entries 必须是数组。" unless entries.is_a?(Array)

      @entries = entries.map { |entry| validate_and_copy(entry) }.freeze
      @counts = @entries.each_with_object(SyncResult.zero) do |entry, result|
        result[ACTION_COUNTS.fetch(entry.action)] += 1
      end.freeze
      freeze
    end

    def to_h
      {
        "entries" => entries.map(&:to_h),
        "counts" => counts.to_h.transform_keys(&:to_s)
      }
    end

    def to_json(*arguments)
      to_h.to_json(*arguments)
    end

    private

    def validate_and_copy(entry)
      raise ArgumentError, "同步计划 entry 类型无效。" unless entry.is_a?(PlanEntry)
      raise ArgumentError, "未知同步计划操作：#{entry.action.inspect}" unless ACTION_COUNTS.key?(entry.action)
      unless entry.name.is_a?(String) && !entry.name.empty?
        raise ArgumentError, "同步计划操作缺少标签名称。"
      end

      validate_desired!(entry)
      validate_reason!(entry)

      PlanEntry.new(
        action: entry.action,
        name: immutable_copy(entry.name),
        desired: immutable_copy(entry.desired),
        reason: entry.reason
      ).freeze
    end

    def validate_desired!(entry)
      if ACTIONS_WITH_DESIRED.include?(entry.action)
        raise ArgumentError, "#{entry.action} 操作缺少目标标签。" unless entry.desired.is_a?(Hash)

        missing = %w[name color description].reject { |key| entry.desired.key?(key) }
        unless missing.empty?
          raise ArgumentError, "#{entry.action} 操作的目标标签缺少字段：#{missing.join(', ')}"
        end
      elsif !entry.desired.nil?
        raise ArgumentError, "#{entry.action} 操作不能包含目标标签。"
      end
    end

    def validate_reason!(entry)
      if entry.action == :delete
        unless DELETE_REASONS.include?(entry.reason)
          raise ArgumentError, "delete 操作的原因无效：#{entry.reason.inspect}"
        end
      elsif !entry.reason.nil?
        raise ArgumentError, "#{entry.action} 操作不能包含删除原因。"
      end
    end

    def immutable_copy(value)
      case value
      when Hash
        value.each_with_object({}) do |(key, item), copy|
          copy[immutable_copy(key)] = immutable_copy(item)
        end.freeze
      when Array
        value.map { |item| immutable_copy(item) }.freeze
      when String
        value.dup.freeze
      else
        value
      end
    end
  end
end

File: src/sync-plan.ts
import type { LabelDefinition } from "./label-types";
import { validatedLabelDefinition } from "./label-definition";
import { OperationCounts, type OperationCountValues } from "./operation-counts";

export type DeleteReason = "legacy_alias" | "stale_managed";

export type PlanEntry =
  | {
      readonly action: "create" | "update" | "rename";
      readonly name: string;
      readonly desired: LabelDefinition;
    }
  | { readonly action: "delete"; readonly name: string; readonly reason: DeleteReason }
  | { readonly action: "unchanged" | "preserve"; readonly name: string };

interface SerializedPlanEntry {
  readonly action: string;
  readonly name: string;
  readonly desired?: LabelDefinition;
  readonly reason?: string;
}

export class SyncPlan {
  readonly entries: readonly PlanEntry[];
  readonly counts: OperationCounts;

  constructor(entries: readonly PlanEntry[]) {
    if (!Array.isArray(entries)) {
      throw new TypeError("同步计划 entries 必须是数组。");
    }

    this.entries = Object.freeze(entries.map((entry) => validateAndCopy(entry)));
    this.counts = OperationCounts.fromOperations(this.entries.map((entry) => entry.action));
    Object.freeze(this);
  }

  toJSON(): {
    readonly entries: readonly SerializedPlanEntry[];
    readonly counts: OperationCountValues;
  } {
    return {
      entries: this.entries.map((entry) => ({ ...entry })),
      counts: this.counts.toJSON(),
    };
  }
}

function validateAndCopy(value: PlanEntry): PlanEntry {
  if (typeof value !== "object" || value === null) {
    throw new TypeError("同步计划 entry 类型无效。");
  }

  const entry = value as PlanEntry;
  if (!["create", "update", "rename", "delete", "unchanged", "preserve"].includes(entry.action)) {
    throw new TypeError(
      `未知同步计划操作：${JSON.stringify((entry as { action?: unknown }).action)}`,
    );
  }
  if (typeof entry.name !== "string" || entry.name.length === 0) {
    throw new TypeError("同步计划操作缺少标签名称。");
  }

  switch (entry.action) {
    case "create":
    case "update":
    case "rename": {
      const desired = immutableLabel(entry.desired, entry.action);
      const reason = (entry as PlanEntry & { readonly reason?: unknown }).reason;
      if (reason !== undefined && reason !== null) {
        throw new TypeError(`${entry.action} 操作不能包含删除原因。`);
      }
      return Object.freeze({
        action: entry.action,
        name: entry.name,
        desired,
      });
    }
    case "delete": {
      const desired = (entry as PlanEntry & { readonly desired?: unknown }).desired;
      if (desired !== undefined && desired !== null) {
        throw new TypeError("delete 操作不能包含目标标签。");
      }
      if (entry.reason !== "legacy_alias" && entry.reason !== "stale_managed") {
        throw new TypeError(`delete 操作的原因无效：${JSON.stringify(entry.reason)}`);
      }
      return Object.freeze({ action: entry.action, name: entry.name, reason: entry.reason });
    }
    case "unchanged":
    case "preserve": {
      const extended = entry as PlanEntry & {
        readonly desired?: unknown;
        readonly reason?: unknown;
      };
      if (extended.desired !== undefined && extended.desired !== null) {
        throw new TypeError(`${entry.action} 操作不能包含目标标签。`);
      }
      if (extended.reason !== undefined && extended.reason !== null) {
        throw new TypeError(`${entry.action} 操作不能包含删除原因。`);
      }
      return Object.freeze({ action: entry.action, name: entry.name });
    }
  }
}

function immutableLabel(
  value: LabelDefinition,
  action: "create" | "update" | "rename",
): LabelDefinition {
  if (typeof value !== "object" || value === null) {
    throw new TypeError(`${action} 操作缺少目标标签。`);
  }
  return validatedLabelDefinition(value, `${action} 操作的目标标签 `);
}
